near-contract-standards/src/access/mod.rs

```diff
@@ -1,129 +1,279 @@
 use std::collections::HashMap;
 
+pub type RoleId = [u8; 32];
+
+/// Contract module which provides a basic access control mechanism, where
+/// there is an account (an [`Self::owner`]) that can be granted exclusive access to
+/// specific functions.
+///
+/// By default, the owner account will be the [one that deploys](env::predecessor_account_id())
+/// the contract. This can later be changed with [`Self::transfer_ownership()`].
+///
+/// This module makes available the modifier [`Self::only_owner()`], which can be applied to your
+///  functions to restrict their use to the owner.
 #[derive(BorshSerialize, BorshDeserialize)]
 pub struct Ownable {
-    pub owner: AccountId,
-    pub default_owner: AccountId,
+    pub owner: Option<AccountId>,
 }
 
 impl Ownable {
+    /// Initializes the contract setting
+    /// the [`predecessor account`](env::predecessor_account_id())
+    /// as the initial [`Self::owner`].
     pub fn new() -> Self {
-        Self {
-            owner: env::predecessor_account_id(),
-            default_owner: "0000000000".parse::<AccountId>().unwrap(),
-        }
+        Self { owner: Some(env::predecessor_account_id()) }
     }
 
-    pub fn owner(&self) -> AccountId {
+    /// Returns the address of the current [`Self::owner`].  
+    /// Panics if it has no owner.
+    pub fn owner(&self) -> Option<AccountId> {
         self.owner.clone()
     }
 
+    /// Has no effect if called by the owner.
+    /// Panics otherwise.
     pub fn only_owner(&self) {
-        require!(env::predecessor_account_id() == self.owner(), "Ownable: caller is not the owner");
+        require!(
+            Some(env::predecessor_account_id()) == self.owner,
+            "Ownable: caller is not the owner"
+        );
     }
 
-    pub fn renounce_ownership(&mut self) {
+    /// Permanently leaves the contract without an [`Self::owner`].  
+    /// Can only be called by the current owner.
+    ///
+    /// # WARNING
+    ///
+    /// Renouncing ownership will leave the contract without an owner,
+    /// thereby removing any functionality that is only available to the owner.  
+    /// ie. future calls into [`Self::only_owner()`] will always panic.
+    pub fn permanently_renounce_ownership(&mut self) {
         self.only_owner();
-        self.owner = self.default_owner.clone();
+        self.owner = None;
     }
 
+    /// Tranfers [ownership](Self::owner) of the contract to a new account `new_owner`.  
+    /// Can only be called by the current owner.
     pub fn transfer_ownership(&mut self, new_owner: AccountId) {
         self.only_owner();
-        require!(new_owner != self.default_owner, "Ownable: new owner is undefined");
-        self.owner = new_owner;
+        self.owner = Some(new_owner);
     }
 }
 
 #[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize)]
 pub struct RoleData {
+    // TODO: consider using HashSet,
+    // in case the bool as false has no particular use
     pub members: HashMap<AccountId, bool>,
```

Comment:
This could possibly be changed into a `HashSet` iff the values of `false` are always equivalent to missing values.

Reply:
You're right. For some reason, I thought that removing the element from the HashMap is more expensive than changing its value, but they are both O(1). Worth changing it to a HashSet for simplicity.

Reply:
Just tried replacing hashmap with hashset, but the build fails. It seems that BorshSerialize and BorshDeserialize are not implemented for hashset.